File: RemoteDesktopClient/app/src/main/java/com/remotedesktop/client/data/Protocol.kt
package com.remotedesktop.client.data

import kotlinx.serialization.Serializable

@Serializable
data class ClientMessage(
    val type: String,
    val pin: String? = null,
    val x: Double? = null,
    val y: Double? = null,
    val dx: Int? = null,
    val dy: Int? = null,
    val button: String? = null,
    val action: String? = null,
    val code: Int? = null,
    val text: String? = null,
    val name: String? = null,
    val quality: Int? = null,
    val scale: Double? = null,
    val fps: Int? = null,
    val timestamp: Long? = null
)

@Serializable
data class ServerResponse(
    val type: String,
    val success: Boolean = false,
    val message: String? = null,
    val serverName: String? = null,
    val screenWidth: Int = 1920,
    val screenHeight: Int = 1080,
    val timestamp: Long = 0
)

@Serializable
data class DiscoveredServer(
    val type: String = "REMOTE_SERVER_INFO",
    val serverName: String,
    val ip: String,
    val port: Int,
    val hasPin: Boolean
)

enum class TouchMode {
    TRACKPAD,
    DIRECT_TOUCH
}

enum class ConnectionState {
    DISCONNECTED,
    CONNECTING,
    AUTHENTICATING,
    CONNECTED,
    ERROR
}

File: RemoteDesktopServer/Network/Protocol.cs
using System.Text.Json.Serialization;

namespace RemoteDesktopServer.Network
{
    public class ClientMessage
    {
        [JsonPropertyName("type")]
        public string Type { get; set; } = string.Empty;

        // Auth
        [JsonPropertyName("pin")]
        public string? Pin { get; set; }

        // Mouse Move Absolute
        [JsonPropertyName("x")]
        public double? X { get; set; }

        [JsonPropertyName("y")]
        public double? Y { get; set; }

        // Mouse Move Delta
        [JsonPropertyName("dx")]
        public int? Dx { get; set; }

        [JsonPropertyName("dy")]
        public int? Dy { get; set; }

        // Mouse Click
        [JsonPropertyName("button")]
        public string? Button { get; set; } // "left", "right", "middle"

        [JsonPropertyName("action")]
        public string? Action { get; set; } // "down", "up", "click", "dblclick", "press"

        // Keyboard & Key code
        [JsonPropertyName("code")]
        public ushort? Code { get; set; }

        [JsonPropertyName("text")]
        public string? Text { get; set; }

        [JsonPropertyName("name")]
        public string? Name { get; set; } // for shortcuts like "win_d", "ctrl_c"

        // Settings / Quality
        [JsonPropertyName("quality")]
        public int? Quality { get; set; }

        [JsonPropertyName("scale")]
        public double? Scale { get; set; }

        [JsonPropertyName("fps")]
        public int? Fps { get; set; }

        // Ping/Pong
        [JsonPropertyName("timestamp")]
        public long? Timestamp { get; set; }
    }

    public class ServerResponse
    {
        [JsonPropertyName("type")]
        public string Type { get; set; } = string.Empty;

        [JsonPropertyName("success")]
        public bool Success { get; set; }

        [JsonPropertyName("message")]
        public string? Message { get; set; }

        [JsonPropertyName("serverName")]
        public string? ServerName { get; set; }

        [JsonPropertyName("screenWidth")]
        public int ScreenWidth { get; set; }

        [JsonPropertyName("screenHeight")]
        public int ScreenHeight { get; set; }

        [JsonPropertyName("timestamp")]
        public long Timestamp { get; set; }
    }
}
